sequencer/src/state.rs

```diff
@@ -1,47 +1,164 @@
-pub struct State {
-    chain_variables: ChainVariables,
-    ethereum_block_tag: BlockNumber,
+#[derive(Hash, Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct ValidatedState {
+    /// Frontier of Block Merkle Tree
+    pub block_merkle_tree: BlockMerkleTree,
+    /// Fee Merkle Tree
+    pub fee_merkle_tree: FeeMerkleTree,
 }
 
-impl HotShotState for State {
-    type Error = Error;
+impl Default for ValidatedState {
+    fn default() -> Self {
+        let block_merkle_tree =
+            BlockMerkleTree::from_elems(Some(32), Vec::<Commitment<Header>>::new()).unwrap();
 
+        // Words of wisdom from @mrain: "capacity = arity^height"
+        // "For index space 2^160, arity 256 (2^8),
+        // you should set the height as 160/8=20"
+        let fee_merkle_tree =
+            FeeMerkleTree::from_kv_set(20, Vec::<(FeeAccount, FeeAmount)>::new()).unwrap();
+        Self {
+            block_merkle_tree,
+            fee_merkle_tree,
+        }
+    }
+}
+
+impl ValidatedState {
+    fn validate_proposal(&self, parent: &Header, proposal: &Header) -> anyhow::Result<Self> {
+        // validate height
+        anyhow::ensure!(
+            proposal.height == parent.height + 1,
+            anyhow::anyhow!(
+                "Invalid Height Error: {}, {}",
+                parent.height,
+                proposal.height
+            )
+        );
+
+        // validate parent fee merkle tree root against state
+        let fee_merkle_tree = self.fee_merkle_tree.clone();
+        let fee_merkle_tree_root = fee_merkle_tree.commitment();
+        anyhow::ensure!(
+            parent.fee_merkle_tree_root == fee_merkle_tree_root,
+            anyhow::anyhow!(
+                "Invalid Fee Merkle Tree Error: {}, {}",
+                parent.fee_merkle_tree_root,
+                fee_merkle_tree_root
+            )
+        );
+
+        // validate parent block merkle tree root against state
+        // `clone` to avoid possible side-effects
+        let mut block_merkle_tree = self.block_merkle_tree.clone();
+        let block_merkle_tree_root = block_merkle_tree.commitment();
+        anyhow::ensure!(
+            parent.block_merkle_tree_root == block_merkle_tree_root,
+            anyhow::anyhow!(
+                "Invalid Block Merkle Tree Error: {}, {}",
+                parent.block_merkle_tree_root,
+                block_merkle_tree_root
+            )
+        );
+
+        // validate proposal is descendent of parent by appending to parent
+        block_merkle_tree.push(parent.commit()).unwrap();
+        let block_merkle_tree_root = block_merkle_tree.commitment();
+        anyhow::ensure!(
+            proposal.block_merkle_tree_root == block_merkle_tree_root,
+            anyhow::anyhow!(
+                "Invalid Block Merkle Tree Error: {}, {}",
+                block_merkle_tree_root,
+                proposal.block_merkle_tree_root
+            )
+        );
+
```

Comment:
We also need to update `fee_merkle_tree` here the same way we do in `Header::new` (e.g. by getting the deposits and inserting them)

Reply:
fixed in 1ad224b1dfae9c120f2505ba5f19fda15f5786a0